src/test/java/io/tiledb/java/api/QueryTest.java

```diff
@@ -173,6 +175,271 @@ public void arrayReadDimensionsTest() throws Exception {
     }
   }
 
+  public static class NIODenseTests {
+
+    @Rule public TemporaryFolder temp = new TemporaryFolder();
+
+    private String arrayURI;
+
+    @Before
+    public void setup() throws Exception {
+      ctx = new Context();
+      arrayURI = temp.getRoot().toPath().resolve("query").toString();
+      arrayCreate();
+      arrayWrite();
+    }
+
+    public void arrayCreate() throws Exception {
+      // The array will be 4x4 with dimensions "rows" and "cols", with domain [1,4].
+      Dimension<Integer> rows =
+          new Dimension<>(ctx, "rows", Integer.class, new Pair<Integer, Integer>(1, 4), 2);
+      Dimension<Integer> cols =
+          new Dimension<>(ctx, "cols", Integer.class, new Pair<Integer, Integer>(1, 4), 2);
+
+      // Create and set getDomain
+      Domain domain = new Domain(ctx);
+      domain.addDimension(rows);
+      domain.addDimension(cols);
+
+      // Add two attributes "a1" and "a2", so each (i,j) cell can store
+      // a character on "a1" and a vector of two floats on "a2".
+      Attribute a1 = new Attribute(ctx, "a1", Character.class);
+      Attribute a2 = new Attribute(ctx, "a2", Float.class);
+      a2.setCellValNum(2);
+
+      ArraySchema schema = new ArraySchema(ctx, TILEDB_DENSE);
+      schema.setTileOrder(TILEDB_ROW_MAJOR);
+      schema.setCellOrder(TILEDB_ROW_MAJOR);
+      schema.setDomain(domain);
+      schema.addAttribute(a1);
+      schema.addAttribute(a2);
+
+      Array.create(arrayURI, schema);
+    }
+
+    public void arrayWrite() throws Exception {
+      String str = "abcdefghijklmnop";
+      float[] floatArr =
+          new float[] {
+            0.1f, 0.2f, 1.1f, 1.2f, 2.1f, 2.2f, 3.1f, 3.2f,
+            4.1f, 4.2f, 5.1f, 5.2f, 6.1f, 6.2f, 7.1f, 7.2f,
+            8.1f, 8.2f, 9.1f, 9.2f, 10.1f, 10.2f, 11.1f, 11.2f,
+            12.1f, 12.2f, 13.1f, 13.2f, 14.1f, 14.2f, 15.1f, 15.2f
+          };
+
+      ByteBuffer a1 = ByteBuffer.allocateDirect(1 * str.length());
+      ByteBuffer a2 = ByteBuffer.allocateDirect(4 * floatArr.length);
+
+      a1.order(ByteOrder.nativeOrder());
+      a2.order(ByteOrder.nativeOrder());
+
+      for (int i = 0; i < floatArr.length; ++i) {
```

Comment:
For the `String`, the following should work:

```
byte[] stringAsBytes = str.getBytes(StandardCharsets.UTF_8);
ByteBuffer a1 = ByteBuffer.allocateDirect(stringAsBytes.length);
a1.put(stringAsBytes);
```

and for the `float[]`:

```
a2.asFloatBuffer().put(floatArr);
```

You can always "upcasting" a `ByteBuffer` in this fashion to get a view of the original buffer in the type of your choosing. Very useful to avoid walking the buffer in a methodwise fashion.